Frameworks/Starboard/EbrStorageFile.cpp

```diff
@@ -0,0 +1,351 @@
+// THE SOFTWARE.
+//
+//******************************************************************************
+
+#include "EbrStorageFile.h"
+#include "ErrorHandling.h"
+
+#include "StringHelpers.h"
+#include "WRLHelpers.h"
+
+#include "RawBuffer.h"
+
+#include <vector>
+#include <fcntl.h>
+
+using namespace Microsoft::WRL;
+using namespace Windows::Foundation;
+using namespace WRLHelpers;
+using namespace ABI::Windows::Storage;
+using namespace ABI::Windows::Storage::AccessCache;
+using namespace ABI::Windows::Storage::FileProperties;
+using namespace ABI::Windows::Storage::Streams;
+
+std::shared_ptr<EbrFile> EbrStorageFile::CreateInstance(const char* path, int mode, int share, int pmode) {
+    // Check to see if this path is a file id path.
+    static const std::string c_fileIdPreamble = "/.file/id=";
+
+    RETURN_NULL_IF_FALSE(0 == strncmp(path, c_fileIdPreamble.c_str() , c_fileIdPreamble.length()));
+
+    Wrappers::HString idHString = Strings::NarrowToWide<HSTRING>(&path[c_fileIdPreamble.length()]);
+
+    ComPtr<IStorageApplicationPermissionsStatics> storageApplicationPermissionsStatics;
+    RETURN_NULL_IF_FAILED(
+        GetActivationFactory(Wrappers::HStringReference(RuntimeClass_Windows_Storage_AccessCache_StorageApplicationPermissions).Get(),
+                             &storageApplicationPermissionsStatics));
+
+    ComPtr<IStorageItemAccessList> accessList;
+    RETURN_NULL_IF_FAILED(storageApplicationPermissionsStatics->get_FutureAccessList(&accessList));
+
+    // Try to grab the item based on id.
+    ComPtr<IStorageFile> storageFile;
+    RETURN_NULL_IF_FAILED(AwaitResult(accessList, &IStorageItemAccessList::GetFileAsync, idHString.Get(), &storageFile));
+
+    bool shouldAllowWrite = ((mode & _O_WRONLY) != 0) ||
+                            ((mode & _O_RDWR) != 0);
+    
+    FileAccessMode openMode = shouldAllowWrite ? FileAccessMode_ReadWrite : FileAccessMode_Read;
+
+    ComPtr<IRandomAccessStream> randomAccessStream;
+    RETURN_NULL_IF_FAILED(AwaitResult(storageFile, &IStorageFile::OpenAsync, openMode, &randomAccessStream));
+
+    bool shouldAppend = ((mode & _O_APPEND) != 0);
+
+    return std::make_shared<EbrStorageFile>(
+        std::move(storageFile),
+        std::move(randomAccessStream),
+        shouldAppend,
+        shouldAllowWrite);
+}
+
+EbrStorageFile::EbrStorageFile(
+        ComPtr<IStorageFile>&& storageFile, 
+        ComPtr<IRandomAccessStream>&& randomAccessStream,
+        bool shouldAppend,
+        bool shouldAllowWrite) : 
+            m_storageFile(storageFile), 
+            m_randomAccessStream(randomAccessStream), 
+            m_shouldAppend(shouldAppend), 
+            m_shouldAllowWrite(shouldAllowWrite) {
+
+    if (shouldAppend) {
+        Lseek(0, SEEK_END);
+    }
+}
+
+int EbrStorageFile::Tell() {
+    return Lseek(0, SEEK_CUR);
+}
+
+int EbrStorageFile::Flush() {
+    std::lock_guard<std::recursive_mutex> lock(m_lock);
+
+    unsigned long long currentPosition{};
```

Comment:
_but why_?

---
In reply to: [84938944](https://github.com/Microsoft/WinObjC/pull/1215#discussion_r84938944) [](ancestors = 84938944)